stdlib/list.gr

```diff
@@ -901,3 +901,1012 @@ provide let sort = (comp, list) => {
+   *
+   * @example List.Mutable.zipWith((a, b) => a + b, fromList([1, 2, 3]), fromList([4, 5, 6])) // fromList([5, 7, 9])
+   * @example List.Mutable.zipWith((a, b) => a * b, fromList([1, 2, 3]), fromList([4, 5])) // fromList([4, 10])
+   *
+   * @since v0.6.0
+   */
+
+  provide let zipWith = (fn, list1, list2) => {
+    fromList(zipWith(fn, toList(list1), toList(list2)))
+  }
+
+  /**
+   * Produces two lists by splitting apart a list of tuples.
+   *
+   * @param list: The list of tuples to split
+   * @returns An list containing all elements from the first tuple element, and a list containing all elements from the second tuple element
+   *
+   * @since v0.6.0
+   */
+
+  provide let unzip = list => {
+    let (list1, list2) = unzip(toList(list))
+    (fromList(list1), fromList(list2))
+  }
+
+  /**
+   * Combines two lists into a Cartesian product of tuples containing
+   * all ordered pairs `(a, b)`.
+   *
+   * @param list1: The list to provide values for the first tuple element
+   * @param list2: The list to provide values for the second tuple element
+   * @returns The new list containing all pairs of `(a, b)`
+   *
+   * @since v0.6.0
+   */
+
+  provide let product = (list1, list2) => {
+    fromList(
+      reduceRight((x1, list) => {
+        reduceRight((x2, list) => [(x1, x2), ...list], list, list2)
+      }, [], list1)
+    )
+  }
+
+  /**
+   * Provides the subset of a list given zero-based start index and amount of elements
+   * to include.
+   *
+   * @param start: The index of the list where the subset will begin (inclusive)
+   * @param length: The amount of elements to be included in the subset
+   * @param list: The input list
+   * @returns The subset of the list
+   *
+   * @throws Failure(String): When `start` is negative
+   * @throws Failure(String): When `length` is negative
+   *
+   * @since v0.6.0
+   */
+
+  provide let sub = (start, length, list) => {
+    fromList(sub(start, length, toList(list)))
+  }
+
+  /**
+   * Reverses the elements in the list.
+   * 
+   * @param list: The list to reverse
+   * 
+   * @since v0.6.0
+   */
+
+  provide let reverse = list => {
+    let rec reverseInner = node => {
+      match (node) {
+        None => void,
+        Some(node) => {
+          let next = node.next
+          node.next = node.prev
+          node.prev = next
+          reverseInner(next)
+        },
+      }
+    }
+    reverseInner(list.head)
+    let head = list.head
+    list.head = list.tail
+    list.tail = head
+  }
+
+  /**
+   * Sorts the list.
+   *
+   * Ordering is calculated using a comparator function which takes two array elements and must return 0 if both are equal, a positive number if the first is greater, and a negative number if the first is smaller.
+   * @param comp: The comparator function used to indicate sort order
+   * @param list: The list to be sorted
+   * 
+   * @since v0.6.0
+   */
+
+  provide let sort = (comp, list) => {
```

Comment:
Do we want this to use the default compare from pervasives maybe with an optional argument? might be a change for a seperate pr.

Reply:
This will have to be rectified in a rebase regardless of whether this PR or the one making `List.sort`'s `comp` a default arg gets merged first so I just left it like this